cranelift/filetests/filetests/isa/x86/windows_fastcall_x64_unwind.clif

```diff
@@ -147,20 +156,53 @@ block0(v0: i64, v1: i64):
     store.i32 v12, v1+80
     store.i32 v13, v1+88
     store.i32 v14, v1+96
+    store.f64 v15, v1+104
+    store.f64 v16, v1+112
+    store.f64 v17, v1+120
+    store.f64 v18, v1+128
+    store.f64 v19, v1+136
+    store.f64 v20, v1+144
+    store.f64 v21, v1+152
+    store.f64 v22, v1+160
+    store.f64 v23, v1+168
     return
 }
 ; sameln: UnwindInfo {
 ; nextln:     version: 1,
 ; nextln:     flags: 0,
-; nextln:     prologue_size: 19,
-; nextln:     unwind_code_count_raw: 10,
+; nextln:     prologue_size: 42,
+; nextln:     unwind_code_count_raw: 16,
 ; nextln:     frame_register: 5,
 ; nextln:     frame_register_offset: 0,
 ; nextln:     unwind_codes: [
 ; nextln:         UnwindCode {
+; nextln:             offset: 42,
+; nextln:             op: SaveXmm128,
+; nextln:             info: 8,
+; nextln:             value: U16(
+; nextln:                 2,
+; nextln:             ),
+; nextln:           },
+; nextln:         UnwindCode {
+; nextln:             offset: 36,
+; nextln:             op: SaveXmm128,
+; nextln:             info: 7,
+; nextln:             value: U16(
+; nextln:                 1,
+; nextln:             ),
+; nextln:         },
+; nextln:         UnwindCode {
+; nextln:             offset: 31,
+; nextln:             op: SaveXmm128,
+; nextln:             info: 6,
+; nextln:             value: U16(
+; nextln:                 0,
```

Comment:
This (and subsequently the above values) seems incorrect.  With an SP adjustment of 0x68, I would expect an offset of 0x60 (8 bytes of alignment to get back to 16 byte alignment required for XMM saves following the pushes) and thus a value of 6, right?

Reply:
My understanding is that offsets are from the lowest address of the function's stack frame. From [MSDN](https://docs.microsoft.com/en-us/cpp/build/exception-handling-x64?view=vs-2019#unwind-operation-code):
> Note: Certain operation codes require an unsigned offset to a value in the local stack frame. This offset is from the start, that is, the lowest address of the fixed stack allocation.

That said, you're absolutely correct that this is just using the offset from `rax`. By happy accident, that yields what I think are the correct offsets - as long as the FPR save region is the last stack slot added, it's probably fine as-is. I'm fixing it to calculate based on the prologue's stack use, so it doesn't rely on happy accidents.

Reply:
So I was wrong because I was comparing our output to VC++, which doesn't use a frame pointer.  Hence I was mistakenly using a positive offset from RSP post-adjustment.  Relevant bit in the docs relating to having a frame pointer:

> If the Frame Register field is nonzero, this offset is from where RSP was located when the FP register was established.  It equals the FP register minus the FP register offset.

I'm not confident that a value of 0 is correct given the above information.  Unless I'm misunderstanding, that would mean `[RBP-0]` would be where it expects MMX6 is being saved, but that's the previous frame pointer.  I would expect perhaps `[RBP-0x40]` (thus a value of 4)?

Reply:
Oh! I think I understand now. "lowest address" must be with respect to the item being offset - lowest address of the 16-byte region `xmm6` would be stored in, for example.

Here's an argument that I hope makes sense, since I'm still trying to work through this myself. Because the offset is calculated from `rsp`/`rbp` (in our case, `rbp` as you noted), the last few entries would be at offset `|GPRs| * 8 + local_alloc_size` and the 16-byte slots before it. This would not include `rbp` since we push that before establishing the frame pointer with a `copy_special`/`mov rbp, rsp`, so this math works out as a `7 * 8 + 0x68 == 0xa0` bytes from `rbp`. Divided by 16 gives an offset of 10 for the furthest, then 9 and 8 for `xmm7` and xmm8` respectively.

I took a quick look at the Windows [unwinder source](https://github.com/dotnet/runtime/blob/e270d9cd31d059ace5b530d592c43cd3594a13f1/src/coreclr/src/unwinder/amd64/unwinder_amd64.cpp#L895-L898) that Jay linked in the earlier incarnation of this PR to see if this would make sense, and I think it's consistent. If `FrameBase` is `rsp` or `rbp` as appropriate then it looks like the offset should just pick the same displacement as the total displacement from the frame register to base of a stack store, minus the offset on the stack store. That comes out to `0xa0` bytes (admittedly the same math - `7 GPRs * 8 bytes + 0x68`) for `xmm6`.

Reply:
So after reading this a bit more, I think I finally understand the offsets for these save ops.  I think the MSDN documentation is worded quite confusingly, although this is a pretty esoteric topic.

In the .NET Core OOM unwinder you linked to, the offsets are relative to `FrameBase` and that is an alias of `EstablisherFrame`.  `EstablisherFrame` is calculated depending upon whether or not a frame pointer is being used in the unwind info.

Without a frame pointer, `FrameBase` ends up being the same thing as RSP:

```
*EstablisherFrame = ContextRecord->Rsp;
```

This is also confirmed with the VC++ unwind information [from this comment](https://github.com/bytecodealliance/cranelift/pull/1378#discussion_r378623302):

```
Prologue:
mov         rax,rsp
sub         rsp,98h
movaps      xmmword ptr [rax-18h],xmm6

Unwind ops:
SAVE_XMM128, register=xmm6 offset=0x80
```

However, when using a frame pointer like we are, `FrameBase` ends up being whatever is stored in the frame pointer *minus* some offset to a "base".

```c
*EstablisherFrame = (&ContextRecord->Rax)[UnwindInfo->FrameRegister];
*EstablisherFrame -= UnwindInfo->FrameOffset * 16;
```

The frame offset is currently being *incorrectly* stored as 0 in our unwind information (the only ops used until now didn't require it, luckily).

So here's what I expect to see.  We should be calculating a "base address" for the offsets, illustrated here with the "base address for unwind offsets":

![](https://www.tortall.net/projects/yasm/manual/html/objfmt-win64/stack-frame-detailed.png)

Note that the base address doesn't have to be after the locals or shadow space like in the illustration (in larger stack allocations it simply can't be; maximum offset is 240).  I'd usually expect it to be at the end of the register save area since that's what the prologue needs to unwind with.  The base offset value simply needs such that `FP - base_offset*16 + op_offset*op_scaling_factor` points at the right thing for all unwind operations with a frame offset.

So does that make sense?  We should have a non-zero "frame offset" value in the root unwind information and then values of 0, 1, and 2 in the XMM save ops would make sense.

Reply:
> The frame offset is currently being incorrectly stored as 0 in our unwind information (the only ops used until now didn't require it, luckily).

I'll have to check in on this too.. Broadly this does make sense, yes. I looked at what I think is [the reference](http://www.tortall.net/projects/yasm/manual/html/manual.html#objfmt-win64-exception) for the diagram you included here - do you know how preparing for a the call to `B` interacts with unwind information? Does Windows require we reserve enough space for the most stack-heavy function call as part of the frame's fixed allocation? (I'm not totally certain we do that, and I'd want to double-check that too)

Reply:
Yes, Windows does require that the local frame's fixed allocation have the maximum amount of space required for passing arguments on the stack for all function calls made by the function, plus 32 bytes for the shadow space which should always be RSP+0x8 to RSP+0x20 upon entry to the callee.

Reply:
Regarding how calling B interacts with the unwind information, it doesn't.  Unwind only concerns itself with unwinding the operations done in the prologue, therefore it is only interested in the register save space and not where any arguments for callees might be stored.  RSP should not adjust (except by a call) after the frame is established unless it's a "dynamic" (alloca) frame, at which point a frame pointer must be used.  Since we're using a frame pointer anyway, we also need that "frame pointer offset" adjustment so that it unwinds from the right base.

What I don't get is why frames with a frame pointer need to be described using this "base address" rather than just treating every unwind operation offset as *negative* from the frame pointer when there is one and positive from RSP when there is not.  This seems unnecessarily complicated, but I'm sure there's a good reason.

Reply:
Great, I think we're on the same page now. I really appreciate your help in figuring this out.

Reply:
I've finally gotten another tweak that I _think_ makes for the right unwind information. Because we always set up `rbp` at the start of the function we're kind of limited to producing functions that have a static frame size of 256 or fewer bytes, and unwinding information is requested. I included [some checks](https://github.com/bytecodealliance/wasmtime/pull/1216/files#diff-890393297114ccac3e3873180274f04aR329-R332) about this, and because we can get away with a wrong offset so long as no unwind codes need it, this can be scoped down to only impacting functions using callee-save FPRs.

This turns what is a (still open) miscompile corrupting arbitrary caller state into a compiler panic, so I think adding a panic here is still a net improvement.

Reply:
Why is the static frame size limited to 256 bytes?  The unwind information should only place limits on the offset from the frame pointer to the "base" (i.e. register save area) and not the entire frame size, no?

Reply:
Since the callee-save region is added after spill slots, it ends up on the lower side of the stack frame, so we need an offset that covers locals and spills, then callee-save FPRs. I fiddled with `stack_layout` a bit to try keeping callee-save FPRs at the higher side of the stack frame (like the diagram shows) but I couldn't figure out how to get that to work.

Reply:
Ugh.  It sounds like we really need to address how we lay out frames in Cranelift if this is a self-imposed limitation.

I mean we can live with this limitation as a short-term stop gap to get the more important issue of not saving the FPRs (i.e. unblock this PR), but I think we should file a high-priority issue to remove this limitation by always saving registers at higher addresses in the frame than any locals and spills.